### PR TITLE
Translate `:gen_statem` termination reports

### PR DESCRIPTION
**Pull request by Moosieus**
This PR addresses an issue where `:gen_statem` modules terminate silently in Elixir, as their termination reports aren't translated. This is particularly troublesome in cases where mix dependencies use `:gen_statem` internally, and may fail silently.

I've made a small repo for reproducing the issue [here](https://github.com/Moosieus/a-sad-state-machine).

I haven't exhaustively evaluated all possible errors or diagnostics that could be included, but I think what's here is a reasonable starting point.

### CODE REVIEW

**Comment by josevalim** [comment]
Thank you. This is a great starting point, the next step is to add tests. You can look at the other tests for examples of how other behaviours are tested.

**Comment by Moosieus** [comment]
I've added tests reflective of what's already in place for `GenServer`, since it's kindred to `:gen_statem`.

I've not put much consideration into the formatting of the `Queue` and `Postponed` lines of the debug output, they just follow the default inspect_opts at the moment.

File: lib/logger/test/logger/translator_test.exs [hunk 1]
```diff
@@ -341,6 +376,180 @@ defmodule Logger.TranslatorTest do
...
+    assert_receive {:event, {:string, ["Process " | _]}, process_metadata}
+
+    assert gen_statem_metadata[:registered_name] == config.test
+    refute Map.has_key?(process_metadata, :registered_name)
+  end
+
+  @tag skip: System.otp_release() < "27"
+  test "translates :gen_statem crashes with process label" do
+    {:ok, pid} = :gen_statem.start(MyGenStatem, :ok)
+    :ok = :gen_statem.call(pid, {:execute, fn -> Process.set_label({:any, "term"}) end})
```
Comment (sabiwara) [inline lib/logger/test/logger/translator_test.exs:427]: This test doesn't seem to work on OTP27:
Screenshot 2024-04-01 at 20 36 18

The implementation also needs to handle `Process Label` like `:gen_event` does for example I think.

Comment (Moosieus) [reply lib/logger/test/logger/translator_test.exs]: I've included the process label when present and fixed that `:gen_statem.start/3` call. All logger tests are now passing for me on OTP 27.0-rc2, let me know if you get the same.

Comment (sabiwara) [reply lib/logger/test/logger/translator_test.exs]: Perfect! Thank you!

**Review by josevalim — APPROVED** [review]

**Comment by whatyouhide** [comment]
Thank you @Moosieus 💟

**Pull request merged**